weaver/cli.py

```diff
@@ -319,6 +319,19 @@ def auth_header(self, token):
         return {"Cookie": token}
 
 
+class SessionAuthHandler(AuthBase):
+    """
+    Adds the cookies from the session to the request.
+    """
+    def __init__(self, session):
+        self.session = session
+
+    def __call__(self, request):
+        # type: (PreparedRequest) -> PreparedRequest
+        request.prepare_cookies(self.session.cookies)
+        return request
```

Comment:
I'm confused. This is exactly what `MagpieAuth` does, which is already supported. 
https://github.com/Ouranosinc/requests-magpie/blob/b26ed3c2cee4395c1bd745c7e102a442ba2199ae/requests_magpie.py#L77-L78
What is different (new use case?) that was not already covered? 
Is it only a matter of `MagpieAuth` providing some sort of `from_session` method? Why must the `session` be used instead of the cookie jar?

Linking back to https://github.com/crim-ca/weaver/pull/597#issuecomment-1930745774, there is no additional request done than the one originally performed when accessing the cookie for the first time since it gets cached in the class (see `self._cookies` in below link). 

[`MagpieAuth`](https://github.com/Ouranosinc/requests-magpie/blob/b26ed3c2cee4395c1bd745c7e102a442ba2199ae/requests_magpie.py#L47-L58) also tries to use a similar interface (on purpose) as [`AuthHandler`](https://github.com/crim-ca/weaver/blob/3a6cc5877950ed651daa779458d151d5adbb4330/weaver/cli.py#L185-L192) so that properties can be employed in the code. This interface is missing many of them.

Reply:
> I'm confused. This is exactly what MagpieAuth does, which is already supported.

MagpieAuth signs in the user with their username and password:
https://github.com/Ouranosinc/requests-magpie/blob/0.2.0/requests_magpie.py#L66

There is no mechanism with MagpieAuth to include cookies that have already been acquired from somewhere else.

> Linking back to https://github.com/crim-ca/weaver/pull/597#issuecomment-1930745774, there is no additional request done than the one originally performed when accessing the cookie for the first time since it gets cached in the class (see self._cookies in below link).

I want a mechanism that doesn't make an original request to acquire the cookies _at all_.  The use-case is that the user already has the cookies, they don't need to get them again.

> [MagpieAuth](https://github.com/Ouranosinc/requests-magpie/blob/b26ed3c2cee4395c1bd745c7e102a442ba2199ae/requests_magpie.py#L47-L58) also tries to use a similar interface (on purpose) as [AuthHandler](https://github.com/crim-ca/weaver/blob/3a6cc5877950ed651daa779458d151d5adbb4330/weaver/cli.py#L185-L192) so that properties can be employed in the code. This interface is missing many of them.

That's fine, I can make this subclass AuthHandler if you'd like but it doesn't share any functionality with AuthHandler really.

Reply:
> I want a mechanism that doesn't make an original request to acquire the cookies at all. The use-case is that the user already has the cookies, they don't need to get them again.

But that would be handled with `CookieAuthHandler` no?
How is the `session` object obtained otherwise? Is it built from the Cookie value? If so, it could be passed directly to avoid this step. 

Note: 
It would be fine with adding a pre-check to `RequestAuthHandler` that skips `request_auth` call when the token/cookie is predefined in the handler if this is what is causing the issue. Ideally, all `AuthHandler` classes should be callable from either Python or the CLI, and I cannot see how to easily achieve the session object propagation in the CLI.

> That's fine, I can make this subclass AuthHandler if you'd like but it doesn't share any functionality with AuthHandler really.

This is fine. It is to be future-proof if username/identity retrieval gets added later on, such as for storing the WPS-outputs under the user-workspace. The method to do it must be defined, but the parameters would at least be available.

Reply:
> But that would be handled with CookieAuthHandler no?

No, CookieAuthHandler still makes a request to obtain the cookie. My use case is: 

>  The use-case is that the user already has the cookies, they don't need to get them again.

> How is the session object obtained otherwise? Is it built from the Cookie value? If so, it could be passed directly to avoid this step.

Yes, that's a good example, it could be built from a cookie value (like the example given in https://github.com/bird-house/birdhouse-deploy/pull/407).
How can it be passed directly? I could build the `Cookie` header and pass that as a custom header to the WeaverClient's methods but it would be a lot nicer to be able to set this once when I initialize the WeaverClient instance.

> It would be fine with adding a pre-check to RequestAuthHandler that skips request_auth call when the token/cookie is predefined in the handler if this is what is causing the issue

Ok I can work with that.